OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -145,240 +52,33 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
 
 		public override object Create(ActorInitializer init) { return new Mobile(init, this); }
 
-		static object LoadSpeeds(MiniYaml y)
-		{
-			var ret = new Dictionary<string, TerrainInfo>();
-			foreach (var t in y.ToDictionary()["TerrainSpeeds"].Nodes)
-			{
-				var speed = FieldLoader.GetValue<int>("speed", t.Value.Value);
-				var nodesDict = t.Value.ToDictionary();
-				var cost = nodesDict.ContainsKey("PathingCost")
-					? FieldLoader.GetValue<int>("cost", nodesDict["PathingCost"].Value)
-					: 10000 / speed;
-				ret.Add(t.Key, new TerrainInfo(speed, cost));
-			}
-
-			return ret;
-		}
-
-		TerrainInfo[] LoadTilesetSpeeds(TileSet tileSet)
-		{
-			var info = new TerrainInfo[tileSet.TerrainInfo.Length];
-			for (var i = 0; i < info.Length; i++)
-				info[i] = TerrainInfo.Impassable;
-
-			foreach (var kvp in TerrainSpeeds)
-			{
-				byte index;
-				if (tileSet.TryGetTerrainIndex(kvp.Key, out index))
-					info[index] = kvp.Value;
-			}
-
-			return info;
-		}
-
-		public class TerrainInfo
-		{
-			public static readonly TerrainInfo Impassable = new TerrainInfo();
-
-			public readonly int Cost;
-			public readonly int Speed;
-
-			public TerrainInfo()
-			{
-				Cost = int.MaxValue;
-				Speed = 0;
-			}
-
-			public TerrainInfo(int speed, int cost)
-			{
-				Speed = speed;
-				Cost = cost;
-			}
-		}
-
-		public struct WorldMovementInfo
-		{
-			internal readonly World World;
-			internal readonly TerrainInfo[] TerrainInfos;
-			internal WorldMovementInfo(World world, MobileInfo info)
-			{
-				// PERF: This struct allows us to cache the terrain info for the tileset used by the world.
-				// This allows us to speed up some performance-sensitive pathfinding calculations.
-				World = world;
-				TerrainInfos = info.TilesetTerrainInfo[world.Map.Rules.TileSet];
-			}
-		}
-
-		public readonly Cache<TileSet, TerrainInfo[]> TilesetTerrainInfo;
-		public readonly Cache<TileSet, int> TilesetMovementClass;
-
-		public MobileInfo()
-		{
-			TilesetTerrainInfo = new Cache<TileSet, TerrainInfo[]>(LoadTilesetSpeeds);
-			TilesetMovementClass = new Cache<TileSet, int>(CalculateTilesetMovementClass);
-		}
-
-		public int MovementCostForCell(World world, CPos cell)
-		{
-			return MovementCostForCell(world, TilesetTerrainInfo[world.Map.Rules.TileSet], cell);
-		}
+		public LocomotorInfo LinkedLocomotorInfo;
```

Comment:
Can we call this just `LocomotorInfo`?  This should also be `{ get; private set; }` to avoid shenanigans.